src/Workspaces/Core/Portable/NamingStyles/EditorConfig/EditorConfigNamingStyleParser.cs

```diff
@@ -0,0 +1,68 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.CodeAnalysis.Diagnostics.Analyzers.NamingStyles
+{
+    internal static partial class EditorConfigNamingStyleParser
+    {
+        /// <remarks>
+        /// The dictionary we get from the VS editorconfig API uses the same dictionary object if there are no changes, so we can cache based on dictionary
+        /// </remarks>
+        private static readonly ConditionalWeakTable<IReadOnlyDictionary<string, object>, SerializableNamingStylePreferencesInfo> _cache = new ConditionalWeakTable<IReadOnlyDictionary<string, object>, SerializableNamingStylePreferencesInfo>();
+        private static readonly object _cacheLock = new object();
+
+        public static SerializableNamingStylePreferencesInfo GetNamingStylesStringFromDictionary(IReadOnlyDictionary<string, object> allRawConventions)
+        {
+            if (!_cache.TryGetValue(allRawConventions, out var value))
```

Comment:
What if we are called on multiple threads and one completes and adds to the cache before we acquire the lock?